pkg/vulnerabilityreport/scanner_test.go

```diff
@@ -1 +1,93 @@
-package vulnerabilityreport_test
+package vulnerabilityreport
+
+import (
+	"github.com/aquasecurity/starboard/pkg/apis/aquasecurity/v1alpha1"
+	"github.com/aquasecurity/starboard/pkg/docker"
+	"github.com/aquasecurity/starboard/pkg/ext"
+	"github.com/aquasecurity/starboard/pkg/kube"
+	"github.com/aquasecurity/starboard/pkg/starboard"
+	"github.com/stretchr/testify/assert"
+	"io"
+	corev1 "k8s.io/api/core/v1"
+	"testing"
+)
+
+//fakePlugin is a minimal implementation of Plugin just sufficient to meet the runtime requirements of a caller of
+//GetScanJobSpec or ParseVulnerabilityScanResult
+type fakePlugin struct{}
+
+func (fakePlugin) GetScanJobSpec(_ corev1.PodSpec, _ map[string]docker.Auth) (corev1.PodSpec, []*corev1.Secret, error) {
+	return corev1.PodSpec{}, []*corev1.Secret{}, nil
+}
+
+func (fakePlugin) ParseVulnerabilityScanResult(_ string, _ io.ReadCloser) (v1alpha1.VulnerabilityScanResult, error) {
+	return v1alpha1.VulnerabilityScanResult{}, nil
+}
+
+func TestPrepareScanJob(t *testing.T) {
+	s := &Scanner{plugin: fakePlugin{}, IDGenerator: ext.NewSimpleIDGenerator()}
+	config := starboard.ConfigData{"vulnerabilityScanner.podTolerations": `[{"key":"k"}]`}
+	job, _, err := s.prepareScanJob(kube.Object{}, corev1.PodSpec{}, map[string]docker.Auth{}, config)
```

Comment:
Testing a private method will test only a fraction of functionality provided by *vulnerabiltyreport.Scanner* type which only exports the *Scan(context.Context, kube.Object)* method. I know that testing the *Scan* method is more challenging but that's something we should probably be doing to allow further refactoring.

Otherwise, each time you change the implementation details (private methods) we'll have to adapt unit tests. Instead of (safe) refactor we'll do code rewrite. WDYT?

Reply:
I agree that testing Scan would be good and I think it would be great to do a PR just focused on that.  In this case, I felt it was beyond the scope of what I was trying to do.

I see what you're saying about testing private methods, but I don't entirely agree.  

Testing a small focused function allows you to more easily and directly exercise its capabilities across  a range of inputs.  And because it's so specific, I find that in practice it's not something that has to change often, so the concern about having to adapt tests each time you change implementation just doesn't come up all that frequently in my experience.  

In contrast, if I wanted to exercise the same 5 cases that I have in TestGetVulnerabilityScannerPodTolerations at the level of a TestScan,  I would have to do a lot of other setup that has nothing to do with the cases of interest.  And those 5 cases would be added to probably 50 or 60 other cases, interested in variations of inputs to the other private methods called by Scan.  

Instead, I'd favor specific tests for specific methods, and then the tests for Scan would only need to prove that the component methods and collaborators were wired up correctly.  In this case, TestPrepareScanJob admittedly leaves most of the details of prepareScanJob untested, but it does prove that getVulnerabilityScannerPodTolerations is wired up.